tests/test_resolver.py

```diff
@@ -1,157 +1,281 @@
+    mock_request.side_effect = Exception('HTTP request failed')
+
+    with pytest.raises(Exception) as exc:
+        Resolver().workspace_from_url(mets_url=url_404)
+
+    # assert
+    assert "HTTP request failed" in str(exc)
+    assert mock_request.call_count == 1
+
+
+def test_workspace_from_url_with_rel_dir(tmp_path):
+    bogus_dst_dir = '../../../../../../../../../../../../../../../../%s' % str(tmp_path)[
+        1:]
+
+    # act
+    with pushd_popd(FOLDER_KANT):
+        ws1 = Resolver().workspace_from_url('data/mets.xml', dst_dir=bogus_dst_dir)
+
+    # assert
+    assert os.path.join(tmp_path, 'mets.xml') == ws1.mets_target
+    assert str(tmp_path) == ws1.directory
+
+
+def test_workspace_from_url0():
+
+    # act
+    workspace = Resolver().workspace_from_url(METS_HEROLD)
+    input_files = workspace.mets.find_all_files(fileGrp='OCR-D-IMG')
+    image_file = input_files[0]
+    f = workspace.download_file(image_file)
+
+    # assert
+    assert '%s.tif' % f.ID == 'FILE_0001_IMAGE.tif'
+    assert f.local_filename == 'OCR-D-IMG/FILE_0001_IMAGE.tif'
+
+
+def test_resolve_image0():
+    workspace = Resolver().workspace_from_url(METS_HEROLD)
+    input_files = workspace.mets.find_all_files(fileGrp='OCR-D-IMG')
+    f = input_files[0]
+    img_pil1 = workspace._resolve_image_as_pil(f.url)
+    assert img_pil1.size == (2875, 3749)
+    img_pil2 = workspace._resolve_image_as_pil(f.url, [[0, 0], [1, 1]])
+    assert img_pil2.size == (1, 1)
+
+
+@pytest.mark.parametrize(
+    "image_url,size1,size2",
+    [('OCR-D-IMG-NRM/OCR-D-IMG-NRM_0017.png', (1457, 2083), (1, 1)),
+     ('OCR-D-IMG-1BIT/OCR-D-IMG-1BIT_0017.png', (1457, 2083), (1, 1)),
+     ])
+def test_resolve_image_grayscale(image_url, size1, size2):
+    url_path = os.path.join(assets.url_of(
+        'kant_aufklaerung_1784-binarized'), 'data/mets.xml')
+    workspace = Resolver().workspace_from_url(url_path)
+    img_url = image_url
+    img_pil1 = workspace.resolve_image_as_pil(img_url)
+    assert img_pil1.size == size1
+    img_pil2 = workspace._resolve_image_as_pil(img_url, [[0, 0], [1, 1]])
+    assert img_pil2.size == size2
+
+
+def test_workspace_from_nothing():
+    ws1 = Resolver().workspace_from_nothing(None)
+    assert ws1.mets
+
+
+def test_workspace_from_nothing_makedirs(tmp_path):
+    non_existant_dir = tmp_path / 'target'
+    ws1 = Resolver().workspace_from_nothing(non_existant_dir)
+    assert ws1.directory == non_existant_dir
+
+
+def test_workspace_from_nothing_noclobber(tmp_path):
+    """Attempt to re-create workspace shall fail because already created
+    """
+
+    ws2 = Resolver().workspace_from_nothing(tmp_path)
+    assert ws2.directory == tmp_path
+
+    with pytest.raises(Exception) as exc:
+        Resolver().workspace_from_nothing(tmp_path)
+
+    # assert
+    the_msg = "METS 'mets.xml' already exists in '%s' and clobber_mets not set" % tmp_path
+    assert the_msg in str(exc)
```

Comment:
Again, `exc` outside its scope... (?)

Reply:
AFAIK it's regular Python Context Management with some extensions. Under the hood it's a Function Call that wraps the actual Expression which raises the Exception and produces an ExceptionInfo-object (the `exc` in the tests mentioned), which carries some of the Exception Attributes to the next lines. The line that raises the Exception must be in any case the final line in the pytest-Context Management, see: https://docs.pytest.org/en/6.2.x/reference.html#pytest.raises

Reply:
Like `The execinfo is not part of the context manager scope (this is true for the context of any context manager), so it lives on after the context manager has exited, making this possible.` see: https://stackoverflow.com/questions/69510562/exception-value-is-not-being-caught-in-pytest-raises-context-manager

Reply:
That's an important detail. So you **must** do the assertion outside pytest.raises. Thanks for clarifying!